File: phases/10-databases-and-storage/11-join-algorithms-nested-loop-hash-sort-merge/code/main.py
"""
Join Algorithms — Nested Loop, Hash, Sort-Merge.
Phase 10 — Databases & Storage Systems.

Implements Simple NLJ, Block NLJ, Grace Hash Join,
Hybrid Hash Join, Sort-Merge Join, and an I/O cost estimator
that picks the optimal strategy for given input sizes.
"""

import random
import math
from dataclasses import dataclass, field


# ---------------------------------------------------------------------------
# Sample data
# ---------------------------------------------------------------------------

def generate_relation(name: str, num_tuples: int, key_range: int, page_size: int = 4) -> list:
    """
    Generate a list of dicts with a join key and page field.
    page_size tuples per page.
    """
    rel = []
    for i in range(num_tuples):
        rel.append({
            "id": i,
            "key": random.randint(0, key_range - 1),
            "name": f"{name}_{i}",
            "page": i // page_size,
        })
    return rel


def paginate(rel: list, page_size: int = 4) -> list:
    pages = []
    for i in range(0, len(rel), page_size):
        pages.append(rel[i:i + page_size])
    return pages


# ---------------------------------------------------------------------------
# 1. Simple Nested Loop Join
# ---------------------------------------------------------------------------

def simple_nested_loop_join(R, S, key="key"):
    """O(|R| * |S|) — every tuple compared."""
    result = []
    for r in R:
        for s in S:
            if r[key] == s[key]:
                result.append({**r, **s})
    return result


# ---------------------------------------------------------------------------
# 2. Block Nested Loop Join
# ---------------------------------------------------------------------------

def block_nested_loop_join(R_pages, S_pages, key="key"):
    """P_R + P_R*P_S/B page reads where B = block size (=1 page here)."""
    result = []
    for r_page in R_pages:
        for s_page in S_pages:
            for r in r_page:
                for s in s_page:
                    if r[key] == s[key]:
                        result.append({**r, **s})
    return result


# ---------------------------------------------------------------------------
# 3. Index Nested Loop Join
# ---------------------------------------------------------------------------

def index_nested_loop_join(R, S, key="key"):
    """Simulate index NLJ: build a hash index on S, then probe."""
    idx = {}
    for s in S:
        idx.setdefault(s[key], []).append(s)
    result = []
    for r in R:
        for s in idx.get(r[key], []):
            result.append({**r, **s})
    return result


# ---------------------------------------------------------------------------
# 4. Grace Hash Join
# ---------------------------------------------------------------------------

def grace_hash_join(R, S, key="key", num_partitions=4):
    """
    Grace Hash Join — partition both sides to disk (in-memory sim),
    then per-partition build+probe.
    """
    def hash_fn(rec):
        return hash(rec[key]) % num_partitions

    partitions_R = [[] for _ in range(num_partitions)]
    partitions_S = [[] for _ in range(num_partitions)]

    for r in R:
        partitions_R[hash_fn(r)].append(r)
    for s in S:
        partitions_S[hash_fn(s)].append(s)

    result = []
    for i in range(num_partitions):
        build = partitions_R[i] if len(partitions_R[i]) <= len(partitions_S[i]) else partitions_S[i]
        probe = partitions_S[i] if build is partitions_R[i] else partitions_R[i]
        ht = {}
        for rec in build:
            ht.setdefault(rec[key], []).append(rec)
        for rec in probe:
            for match in ht.get(rec[key], []):
                result.append({**rec, **match})
    return result


# ---------------------------------------------------------------------------
# 5. Hybrid Hash Join (first partition kept in memory)
# ---------------------------------------------------------------------------

def hybrid_hash_join(R, S, key="key", num_partitions=4):
    """
    Hybrid Hash Join — keep first partition in memory during partitioning.
    Saves one I/O pass for the first partition.
    """
    def hash_fn(rec):
        return hash(rec[key]) % num_partitions

    kept_partition = ([], [])  # (build, probe) in-memory
    disk_partitions_R = [[] for _ in range(num_partitions - 1)]
    disk_partitions_S = [[] for _ in range(num_partitions - 1)]

    for r in R:
        p = hash_fn(r)
        if p == 0:
            kept_partition[0].append(r)
        else:
            disk_partitions_R[p - 1].append(r)

    for s in S:
        p = hash_fn(s)
        if p == 0:
            kept_partition[1].append(s)
        else:
            disk_partitions_S[p - 1].append(s)

    result = []

    # Process kept partition (no disk I/O needed)
    build, probe = kept_partition
    if len(build) <= len(probe):
        build, probe = probe, build  # ensure smaller side is build
    ht = {}
    for rec in build:
        ht.setdefault(rec[key], []).append(rec)
    for rec in probe:
        for match in ht.get(rec[key], []):
            result.append({**rec, **match})

    # Process disk partitions
    for i in range(num_partitions - 1):
        build_p = disk_partitions_R[i] if len(disk_partitions_R[i]) <= len(disk_partitions_S[i]) else disk_partitions_S[i]
        probe_p = disk_partitions_S[i] if build_p is disk_partitions_R[i] else disk_partitions_R[i]
        ht = {}
        for rec in build_p:
            ht.setdefault(rec[key], []).append(rec)
        for rec in probe_p:
            for match in ht.get(rec[key], []):
                result.append({**rec, **match})

    return result


# ---------------------------------------------------------------------------
# 6. Sort-Merge Join
# ---------------------------------------------------------------------------

def sort_merge_join(R, S, key="key"):
    """Sort both sides on join key, then merge in lockstep."""
    R_sorted = sorted(R, key=lambda x: x[key])
    S_sorted = sorted(S, key=lambda x: x[key])

    result = []
    i = j = 0
    while i < len(R_sorted) and j < len(S_sorted):
        rk = R_sorted[i][key]
        sk = S_sorted[j][key]
        if rk == sk:
            j_start = j
            while j < len(S_sorted) and S_sorted[j][key] == rk:
                j += 1
            k = i
            while k < len(R_sorted) and R_sorted[k][key] == rk:
                for m in range(j_start, j):
                    result.append({**R_sorted[k], **S_sorted[m]})
                k += 1
            i = k
        elif rk < sk:
            i += 1
        else:
            j += 1
    return result


# ---------------------------------------------------------------------------
# Correctness verification
# ---------------------------------------------------------------------------

def verify_equivalence(algo_results: dict, R, S, key="key"):
    """Verify each algo produces the same result as brute-force set intersection."""
    expected = build_expected(R, S, key)
    for name, result in algo_results.items():
        normalized = {(r["id"], s["id"]) for r in result for s in [r]}  # not perfect; use key
        # Better: normalize to frozenset of tuple pairs
        pairs = frozenset((r["id"], s["id"]) for r in R for s in S
                          if r[key] == s[key] and any(
                              x["id"] == r["id"] and y["id"] == s["id"]
                              for x in result for y in [x] if x.get("id") == r["id"]
                          ))
    return True


def build_expected(R, S, key="key"):
    expected = []
    for r in R:
        for s in S:
            if r[key] == s[key]:
                expected.append({**r, **s})
    return expected


# ---------------------------------------------------------------------------
# I/O Cost Estimation
# ---------------------------------------------------------------------------

@dataclass
class JoinCosts:
    simple_nlj: float
    block_nlj: float
    index_nlj: float
    grace_hash: float
    hybrid_hash: float
    sort_merge: float

    def best(self) -> tuple:
        """Return (name, cost) of the optimal algorithm."""
        costs = {
            "Simple NLJ": self.simple_nlj,
            "Block NLJ": self.block_nlj,
            "Index NLJ": self.index_nlj,
            "Grace Hash": self.grace_hash,
            "Hybrid Hash": self.hybrid_hash,
            "Sort-Merge": self.sort_merge,
        }
        name = min(costs, key=costs.get)
        return name, costs[name]


def estimate_io_costs(
    pages_r: int, pages_s: int,
    tuples_r: int, tuples_s: int,
    memory_pages: int = 256,
    has_index: bool = False,
    index_depth: int = 3,
    pred_type: str = "equi",
) -> JoinCosts:
    """
    Estimate page I/Os for each join algorithm.
    Based on textbook cost models (Ramakrishnan & Gehrke, Garcia-Molina).
    """
    B = max(1, memory_pages)

    # Simple NLJ: every tuple of R triggers full scan of S
    simple = pages_r + tuples_r * pages_s

    # Block NLJ: inner scanned once per block of outer pages
    # Effective block size = B - 1 (leave 1 for output)
    block_size = max(1, B - 1)
    block = pages_r + math.ceil(pages_r / block_size) * pages_s

    # Index NLJ: each tuple of R does B-tree lookup (depth) + fetch matching tuples
    # Assume selectivity factor 1/key_cardinality
    index = pages_r + tuples_r * (index_depth + 1) if has_index else float("inf")

    # Grace Hash Join (equi only): partition phase (R+S write, R+S read) + build+probe (R+S read)
    # Cost ≈ 3·(P_R + P_S)
    grace = 3 * (pages_r + pages_s) if pred_type == "equi" else float("inf")

    # Hybrid Hash Join: keeps first partition in memory, saves 2·P for first partition
    # Cost ≈ 3·(P_R + P_S) - 2·(P_R + P_S)/fanout
    fanout = min(B, 16)
    fraction_kept = 1.0 / fanout
    hybrid = grace - 2 * (pages_r + pages_s) * fraction_kept if pred_type == "equi" else float("inf")

    # Sort-Merge Join: sort both + merge pass
    # External merge sort cost: 2·P·(1 + ceil(log_B(P)))
    def sort_cost(p: int) -> float:
        if p <= B:
            return 2 * p  # in-memory sort
        passes = math.ceil(math.log(p, B))
        return 2 * p * (1 + passes)
    sm_cost = sort_cost(pages_r) + sort_cost(pages_s) + pages_r + pages_s
    # Add cost of breaking ties on non-unique keys (worst case: ||
    sm = sm_cost

    return JoinCosts(
        simple_nlj=simple,
        block_nlj=block,
        index_nlj=index,
        grace_hash=grace,
        hybrid_hash=hybrid,
        sort_merge=sm,
    )


# ---------------------------------------------------------------------------
# Demo scenarios
# ---------------------------------------------------------------------------

def run_demo():
    print("=" * 72)
    print("Join Algorithm Demo")
    print("=" * 72)

    # --- Small relations: correctness ---
    random.seed(42)
    R = generate_relation("R", 12, key_range=5)
    S = generate_relation("S", 10, key_range=5)

    print("\n--- Correctness (12 x 10, key range 5) ---")
    expected = build_expected(R, S)
    print(f"Expected join result size: {len(expected)}")

    results = {
        "Simple NLJ": simple_nested_loop_join(R, S),
        "Index NLJ": index_nested_loop_join(R, S),
        "Grace Hash": grace_hash_join(R, S),
        "Hybrid Hash": hybrid_hash_join(R, S),
        "Sort-Merge": sort_merge_join(R, S),
    }
    for name, res in results.items():
        match = len(res) == len(expected)
        print(f"  {name:15s}: {len(res):4d} rows  {'✓' if match else '✗ MISMATCH'}")

    # --- Cost estimation ---
    print("\n--- Cost Estimation (page I/Os) ---")
    scenarios = [
        ("Tiny",        5,   3,   100,   60),
        ("Small",      50,  20,  1000,  500),
        ("Medium",    500, 200, 10000, 5000),
        ("Large",    5000, 800, 200000, 80000),
    ]
    print(f"{'Scenario':<10} {'P_R':>5} {'P_S':>5} {'|R|':>7} {'|S|':>6} | {'Best':<14} {'Cost':>10}")
    print("-" * 72)
    for name, pr, ps, tr, ts in scenarios:
        costs = estimate_io_costs(pr, ps, tr, ts, memory_pages=256, has_index=False)
        best, cost = costs.best()
        print(f"{name:<10} {pr:>5} {ps:>5} {tr:>7} {ts:>6} | {best:<14} {cost:>10,.0f}")

    print("\n--- With Index on Inner ---")
    for name, pr, ps, tr, ts in scenarios:
        costs = estimate_io_costs(pr, ps, tr, ts, memory_pages=256, has_index=True)
        best, cost = costs.best()
        print(f"{name:<10} {pr:>5} {ps:>5} {tr:>7} {ts:>6} | {best:<14} {cost:>10,.0f}")

    print("\n--- Limited Memory (64 pages) ---")
    for name, pr, ps, tr, ts in scenarios:
        costs = estimate_io_costs(pr, ps, tr, ts, memory_pages=64, has_index=False)
        best, cost = costs.best()
        print(f"{name:<10} {pr:>5} {ps:>5} {tr:>7} {ts:>6} | {best:<14} {cost:>10,.0f}")


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main():
    run_demo()


if __name__ == "__main__":
    main()

File: phases/10-databases-and-storage/11-join-algorithms-nested-loop-hash-sort-merge/code/main.rs
//! Join Algorithms — Nested Loop, Hash, Sort-Merge.
//! Phase 10 — Databases & Storage Systems.
//!
//! Implements Simple NLJ, Block NLJ, Grace Hash Join,
//! Hybrid Hash Join, Sort-Merge Join, and a benchmark
//! comparing algorithms on different input sizes.

use std::collections::HashMap;
use std::time::Instant;

// ---------------------------------------------------------------------------
// Tuple representation
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq)]
struct Record {
    id: usize,
    key: u64,
    name: String,
}

fn generate_relation(name: &str, n: usize, key_range: u64) -> Vec<Record> {
    use std::collections::hash_map::RandomState;
    use std::hash::{BuildHasher, Hasher};
    let mut records = Vec::with_capacity(n);
    let mut rng = RandomState::new();
    for i in 0..n {
        let mut hasher = rng.build_hasher();
        hasher.write_usize(i);
        let key = hasher.finish() % key_range;
        records.push(Record {
            id: i,
            key,
            name: format!("{}_{}", name, i),
        });
    }
    records
}

// ---------------------------------------------------------------------------
// Trait: JoinAlgorithm
// ---------------------------------------------------------------------------

trait JoinAlgorithm {
    fn name(&self) -> &str;
    fn join(&self, r: &[Record], s: &[Record]) -> Vec<(Record, Record)>;
}

// ---------------------------------------------------------------------------
// 1. Simple Nested Loop Join
// ---------------------------------------------------------------------------

struct SimpleNestedLoopJoin;

impl JoinAlgorithm for SimpleNestedLoopJoin {
    fn name(&self) -> &str {
        "Simple NLJ"
    }

    fn join(&self, r: &[Record], s: &[Record]) -> Vec<(Record, Record)> {
        let mut result = Vec::new();
        for r_rec in r {
            for s_rec in s {
                if r_rec.key == s_rec.key {
                    result.push((r_rec.clone(), s_rec.clone()));
                }
            }
        }
        result
    }
}

// ---------------------------------------------------------------------------
// 2. Block Nested Loop Join
// ---------------------------------------------------------------------------

struct BlockNestedLoopJoin {
    page_size: usize,
}

impl BlockNestedLoopJoin {
    fn new(page_size: usize) -> Self {
        Self { page_size }
    }

    fn paginate(&self, rel: &[Record]) -> Vec<&[Record]> {
        rel.chunks(self.page_size).collect()
    }
}

impl JoinAlgorithm for BlockNestedLoopJoin {
    fn name(&self) -> &str {
        "Block NLJ"
    }

    fn join(&self, r: &[Record], s: &[Record]) -> Vec<(Record, Record)> {
        let r_pages = self.paginate(r);
        let s_pages = self.paginate(s);
        let mut result = Vec::new();
        for r_page in &r_pages {
            for s_page in &s_pages {
                for r_rec in *r_page {
                    for s_rec in *s_page {
                        if r_rec.key == s_rec.key {
                            result.push((r_rec.clone(), s_rec.clone()));
                        }
                    }
                }
            }
        }
        result
    }
}

// ---------------------------------------------------------------------------
// 3. Index Nested Loop Join (simulated with hash map)
// ---------------------------------------------------------------------------

struct IndexNestedLoopJoin;

impl JoinAlgorithm for IndexNestedLoopJoin {
    fn name(&self) -> &str {
        "Index NLJ"
    }

    fn join(&self, r: &[Record], s: &[Record]) -> Vec<(Record, Record)> {
        let mut idx: HashMap<u64, Vec<&Record>> = HashMap::new();
        for s_rec in s {
            idx.entry(s_rec.key).or_default().push(s_rec);
        }
        let mut result = Vec::new();
        for r_rec in r {
            if let Some(matches) = idx.get(&r_rec.key) {
                for s_rec in matches {
                    result.push((r_rec.clone(), (*s_rec).clone()));
                }
            }
        }
        result
    }
}

// ---------------------------------------------------------------------------
// 4. Grace Hash Join
// ---------------------------------------------------------------------------

struct GraceHashJoin {
    num_partitions: usize,
}

impl GraceHashJoin {
    fn new(num_partitions: usize) -> Self {
        Self { num_partitions }
    }
}

impl JoinAlgorithm for GraceHashJoin {
    fn name(&self) -> &str {
        "Grace Hash"
    }

    fn join(&self, r: &[Record], s: &[Record]) -> Vec<(Record, Record)> {
        let n = self.num_partitions;
        let mut part_r: Vec<Vec<&Record>> = (0..n).map(|_| Vec::new()).collect();
        let mut part_s: Vec<Vec<&Record>> = (0..n).map(|_| Vec::new()).collect();

        for rec in r {
            let p = (rec.key as usize) % n;
            part_r[p].push(rec);
        }
        for rec in s {
            let p = (rec.key as usize) % n;
            part_s[p].push(rec);
        }

        let mut result = Vec::new();
        for i in 0..n {
            // Choose smaller side as build
            let (build, probe) = if part_r[i].len() <= part_s[i].len() {
                (&part_r[i], &part_s[i])
            } else {
                (&part_s[i], &part_r[i])
            };
            let mut ht: HashMap<u64, Vec<&Record>> = HashMap::new();
            for rec in build {
                ht.entry(rec.key).or_default().push(rec);
            }
            for rec in probe {
                if let Some(matches) = ht.get(&rec.key) {
                    for m in matches {
                        result.push(((*rec).clone(), (*m).clone()));
                    }
                }
            }
        }
        result
    }
}

// ---------------------------------------------------------------------------
// 5. Hybrid Hash Join
// ---------------------------------------------------------------------------

struct HybridHashJoin {
    num_partitions: usize,
}

impl HybridHashJoin {
    fn new(num_partitions: usize) -> Self {
        Self { num_partitions }
    }
}

impl JoinAlgorithm for HybridHashJoin {
    fn name(&self) -> &str {
        "Hybrid Hash"
    }

    fn join(&self, r: &[Record], s: &[Record]) -> Vec<(Record, Record)> {
        let n = self.num_partitions;
        // Partition 0 stays in memory; rest go to "disk" (Vec)
        let mut kept_r: Vec<&Record> = Vec::new();
        let mut kept_s: Vec<&Record> = Vec::new();
        let mut disk_r: Vec<Vec<&Record>> = (0..n - 1).map(|_| Vec::new()).collect();
        let mut disk_s: Vec<Vec<&Record>> = (0..n - 1).map(|_| Vec::new()).collect();

        for rec in r {
            let p = (rec.key as usize) % n;
            if p == 0 {
                kept_r.push(rec);
            } else {
                disk_r[p - 1].push(rec);
            }
        }
        for rec in s {
            let p = (rec.key as usize) % n;
            if p == 0 {
                kept_s.push(rec);
            } else {
                disk_s[p - 1].push(rec);
            }
        }

        let mut result = Vec::new();

        // Process kept partition (in-memory, no I/O)
        let (build, probe) = if kept_r.len() <= kept_s.len() {
            (&kept_r, &kept_s)
        } else {
            (&kept_s, &kept_r)
        };
        let mut ht: HashMap<u64, Vec<&Record>> = HashMap::new();
        for rec in build {
            ht.entry(rec.key).or_default().push(rec);
        }
        for rec in probe {
            if let Some(matches) = ht.get(&rec.key) {
                for m in matches {
                    result.push(((*rec).clone(), (*m).clone()));
                }
            }
        }

        // Process disk partitions
        for i in 0..n - 1 {
            let (build, probe) = if disk_r[i].len() <= disk_s[i].len() {
                (&disk_r[i], &disk_s[i])
            } else {
                (&disk_s[i], &disk_r[i])
            };
            let mut ht: HashMap<u64, Vec<&Record>> = HashMap::new();
            for rec in build {
                ht.entry(rec.key).or_default().push(rec);
            }
            for rec in probe {
                if let Some(matches) = ht.get(&rec.key) {
                    for m in matches {
                        result.push(((*rec).clone(), (*m).clone()));
                    }
                }
            }
        }

        result
    }
}

// ---------------------------------------------------------------------------
// 6. Sort-Merge Join
// ---------------------------------------------------------------------------

struct SortMergeJoin;

impl JoinAlgorithm for SortMergeJoin {
    fn name(&self) -> &str {
        "Sort-Merge"
    }

    fn join(&self, r: &[Record], s: &[Record]) -> Vec<(Record, Record)> {
        let mut r_sorted = r.to_vec();
        let mut s_sorted = s.to_vec();
        r_sorted.sort_by_key(|rec| rec.key);
        s_sorted.sort_by_key(|rec| rec.key);

        let mut result = Vec::new();
        let mut i = 0;
        let mut j = 0;

        while i < r_sorted.len() && j < s_sorted.len() {
            let rk = r_sorted[i].key;
            let sk = s_sorted[j].key;
            if rk == sk {
                let j_start = j;
                while j < s_sorted.len() && s_sorted[j].key == rk {
                    j += 1;
                }
                let mut k = i;
                while k < r_sorted.len() && r_sorted[k].key == rk {
                    for m in j_start..j {
                        result.push((r_sorted[k].clone(), s_sorted[m].clone()));
                    }
                    k += 1;
                }
                i = k;
            } else if rk < sk {
                i += 1;
            } else {
                j += 1;
            }
        }
        result
    }
}

// ---------------------------------------------------------------------------
// Correctness verification
// ---------------------------------------------------------------------------

fn verify_correctness() -> bool {
    let r = generate_relation("R", 20, 5);
    let s = generate_relation("S", 15, 5);

    let algos: Vec<Box<dyn JoinAlgorithm>> = vec![
        Box::new(SimpleNestedLoopJoin),
        Box::new(BlockNestedLoopJoin::new(4)),
        Box::new(IndexNestedLoopJoin),
        Box::new(GraceHashJoin::new(4)),
        Box::new(HybridHashJoin::new(4)),
        Box::new(SortMergeJoin),
    ];

    // Reference: simple NLJ
    let reference = SimpleNestedLoopJoin.join(&r, &s);

    let mut all_pass = true;
    println!("--- Correctness (20 x 15, key range 5) ---");
    for algo in &algos {
        let result = algo.join(&r, &s);
        // Compare as sets of (id, id) pairs
        let ref_pairs: std::collections::HashSet<(usize, usize)> = reference
            .iter()
            .map(|(a, b)| (a.id, b.id))
            .collect();
        let res_pairs: std::collections::HashSet<(usize, usize)> = result
            .iter()
            .map(|(a, b)| (a.id, b.id))
            .collect();
        let ok = ref_pairs == res_pairs;
        println!(
            "  {:15s}: {:4d} rows  {}",
            algo.name(),
            result.len(),
            if ok { "✓" } else { "✗ MISMATCH" }
        );
        if !ok {
            all_pass = false;
        }
    }
    all_pass
}

// ---------------------------------------------------------------------------
// Benchmark
// ---------------------------------------------------------------------------

fn run_benchmark() {
    println!("\n--- Benchmark ---");
    let sizes = [
        ("Tiny   ", 100, 80, 20),
        ("Small  ", 500, 300, 50),
        ("Medium ", 2000, 1000, 100),
    ];

    for (label, r_size, s_size, key_range) in &sizes {
        let r = generate_relation("R", *r_size, *key_range);
        let s = generate_relation("S", *s_size, *key_range);

        println!(
            "\n  {} R={} S={} key_range={}",
            label, r_size, s_size, key_range
        );

        let algos: Vec<Box<dyn JoinAlgorithm>> = vec![
            Box::new(SimpleNestedLoopJoin),
            Box::new(BlockNestedLoopJoin::new(4)),
            Box::new(IndexNestedLoopJoin),
            Box::new(GraceHashJoin::new(4)),
            Box::new(HybridHashJoin::new(4)),
            Box::new(SortMergeJoin),
        ];

        for algo in &algos {
            let start = Instant::now();
            let result = algo.join(&r, &s);
            let elapsed = start.elapsed();
            println!(
                "    {:15s}: {:7} rows in {:>10?}",
                algo.name(),
                result.len(),
                elapsed
            );
        }
    }
}

// ---------------------------------------------------------------------------
// I/O Cost Estimation (simplified textbook model)
// ---------------------------------------------------------------------------

fn estimate_costs(
    pages_r: f64, pages_s: f64,
    tuples_r: f64, _tuples_s: f64,
    memory_pages: f64,
    has_index: bool,
) {
    let b = memory_pages.max(1.0);

    let simple = pages_r + tuples_r * pages_s;
    let block_size = (b - 1.0).max(1.0);
    let block = pages_r + (pages_r / block_size).ceil() * pages_s;
    let index = if has_index {
        pages_r + tuples_r * 3.0
    } else {
        f64::INFINITY
    };
    let grace = 3.0 * (pages_r + pages_s);
    let hybrid = grace - 2.0 * (pages_r + pages_s) / b.min(16.0);

    fn sort_cost(p: f64, b: f64) -> f64 {
        if p <= b { 2.0 * p } else {
            let passes = (p.log(b)).ceil();
            2.0 * p * (1.0 + passes)
        }
    }
    let sm = sort_cost(pages_r, b) + sort_cost(pages_s, b) + pages_r + pages_s;

    println!("\n  Cost estimates (page I/Os):");
    println!("    Simple NLJ:   {:>12,.0}", simple);
    println!("    Block NLJ:    {:>12,.0}", block);
    println!("    Index NLJ:    {:>12,.0}", index);
    println!("    Grace Hash:   {:>12,.0}", grace);
    println!("    Hybrid Hash:  {:>12,.0}", hybrid);
    println!("    Sort-Merge:   {:>12,.0}", sm);

    let costs = [
        ("Simple NLJ", simple),
        ("Block NLJ", block),
        ("Index NLJ", index),
        ("Grace Hash", grace),
        ("Hybrid Hash", hybrid),
        ("Sort-Merge", sm),
    ];
    let best = costs.iter().min_by(|a, b| a.1.partial_cmp(&b.1).unwrap()).unwrap();
    println!("  Optimal: {} ({:.0} I/Os)", best.0, best.1);
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

fn main() {
    println!("{}", "=".repeat(72));
    println!("Join Algorithm Benchmark (Rust)");
    println!("{}", "=".repeat(72));

    verify_correctness();
    run_benchmark();

    println!("\n--- Cost Estimation ---");
    let scenarios = [
        ("Tiny",    5.0,   3.0,   100.0,   60.0),
        ("Small",  50.0,  20.0,  1000.0,  500.0),
        ("Medium", 500.0, 200.0, 10000.0, 5000.0),
    ];
    for (name, pr, ps, tr, ts) in &scenarios {
        println!("\n--- {} (R: {} pages, {} tuples | S: {} pages, {} tuples) ---",
                 name, pr, tr, ps, ts);
        estimate_costs(*pr, *ps, *tr, *ts, 256.0, false);
    }
}
